src/status.rs

```diff
@@ -309,9 +427,370 @@ mod test {
+
+    // is_disconnected()
+    #[test]
+    fn no_recent_hbs_on_both_sides() {
+        let server = SERVER_POOL.get_server();
+        let channel = get_test_channel(&server);
+        let heartbeat_1 = get_heartbeat_msgs(Duration::minutes(10));
+        let heartbeat_2 = get_heartbeat_msgs(Duration::minutes(10));
+        let new_state = get_newstate_msg();
+        let approve_state = get_approve_state_msg();
+
+        let messages = Messages {
+            leader: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: Some(new_state),
+                    approve_state: None,
+                }),
+                heartbeats: Some(vec![heartbeat_1]),
+            },
+            follower: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: None,
+                    approve_state: Some(approve_state),
+                }),
+                heartbeats: Some(vec![heartbeat_2]),
+            },
+            recency: Duration::minutes(4),
+        };
+
+        assert_eq!(
+            is_disconnected(&channel, &messages),
+            true,
+            "Both leader and follower heartbeats have no recent messages"
+        )
+    }
+
+    #[test]
+    fn no_recent_follower_hbs() {
+        let server = SERVER_POOL.get_server();
+        let channel = get_test_channel(&server);
+        let mut heartbeat_1 = get_heartbeat_msgs(Duration::minutes(10)); // not recent
+        heartbeat_1.from = channel.spec.validators.leader().id;
+        let mut heartbeat_2 = get_heartbeat_msgs(Duration::minutes(10));
+        heartbeat_2.from = channel.spec.validators.follower().id;
+        let mut heartbeat_3 = get_heartbeat_msgs(Duration::minutes(10));
+        heartbeat_3.from = channel.spec.validators.leader().id;
+        let mut heartbeat_4 = get_heartbeat_msgs(Duration::minutes(10));
+        heartbeat_4.from = channel.spec.validators.follower().id;
+        let new_state = get_newstate_msg();
+        let approve_state = get_approve_state_msg();
+
+        let messages = Messages {
+            leader: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: Some(new_state),
+                    approve_state: None,
+                }),
+                heartbeats: Some(vec![heartbeat_1, heartbeat_2]),
+            },
+            follower: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: None,
+                    approve_state: Some(approve_state),
+                }),
+                heartbeats: Some(vec![heartbeat_3, heartbeat_4]),
+            },
+            recency: Duration::minutes(4),
+        };
+
+        assert_eq!(
+            is_disconnected(&channel, &messages),
+            true,
+            "No recent heartbeats on both validators"
+        )
+    }
+
+    #[test]
+    fn no_recent_leader_hb_coming_from_follower() {
```

Comment:
maybe the name is confusing me, but don't we have a Recent HB with `from: leader` in the follower, i.e. `heartbeat_2`?
If it's only the naming, maybe we can rename it to something more clear?